cstar/base/utils.py

```
@@ -338,3 +340,87 @@ def _dict_to_tree(input_dict: dict, prefix: str = "") -> str:
+    msg_post (str | None), default = None):
+        An overridden message logged after the command is successfully executed.
+    msg_err (str | None), default = None):
+        An overridden message logged when a command returns a non-zero code. Logs
+        will be supplied with the subprocess output using the name `result`,
+        e.g. `msg_err.format(result=result)`.
+    raise_on_error (bool, default = False):
+        If True, raises a RuntimeError if the command returns a non-zero code.
+
+    Returns:
+    -------
+    stdout: str
+       The captured standard output of the process.
+
+    Examples:
+    --------
+    In: _run_cmd("python foo.py", "Running script", "Script completed")
+    Out: Running script
+         Script completed
+
+    In: _run_cmd("python foo.py")
+    Out: Running command: python foo.py
+         Command completed successfully. STDOUT: <output of foo.py>
+
+    In: _run_cmd("python return_nonzero.py")
+    Out: Running command: python return_nonzero.py
+         Command python return_nonzero.py failed: <stderror of foo.py>
+    """
+    print(msg_pre or f"Running command: {cmd}")
+    stdout: str = ""
+
+    fn = functools.partial(
+        subprocess.run,
+        cmd,
+        shell=True,
+        text=True,
+        capture_output=True,
+    )
+
+    kwargs: dict[str, str | PathLike | dict[str, str]] = {}
+    if cwd:
+        kwargs["cwd"] = cwd
+    if env:
+        kwargs["env"] = env
+
+    result: subprocess.CompletedProcess[str] = fn(**kwargs)
+    stdout = str(result.stdout).strip() if result.stdout is not None else ""
+
+    if result.returncode != 0:
+        if msg_err:
+            msg = msg_err.format(result=result)
+        else:
+            msg = f"Command `{cmd}` failed: {result.stderr.strip()}"
+
+        if raise_on_error:
+            raise RuntimeError(msg)
+
+        print(msg)
+
+    print(msg_post or f"Command completed successfully. STDOUT: {result.stdout}")
```

As with my above comment re. a default print statement, I would suggest against printing STDOUT as the default if nothing has gone wrong, we're hoping to hide all the internals of most things from the user, but again this is probably addressed in the `logging` PR. I was imagining something like logging all these things somewhere an advanced user could go looking for them

---

Adjusted existing success msg.